File: cmd/admin/rename.py
import logging
import re
import unicodedata

import discord
from discord.ext import commands

from utils.core.embeds import make_embed
from utils.core.emojis import EMOJIS
from utils.logging.mod_log import send_mod_log
from utils.permissions.base_admin import BaseAdminCog

logger = logging.getLogger("bot")

PROFANITY_BLOCKLIST = {
    "madarchod", "bhenchod", "behenchod", "chutiya", "chut", "gaand", "gandu",
    "randi", "kutti", "kutta", "lund", "lavda", "loda", "harami"
}


class RenameSystem(BaseAdminCog):
    """Cog responsible for server member nickname management and profanity filtering."""

    COOLDOWN_RATE = 1
    COOLDOWN_PER = 5

    def __init__(self, bot: commands.Bot) -> None:
        self.bot = bot

    async def _delete_command(self, ctx: commands.Context) -> None:
        try:
            await ctx.message.delete()
        except (discord.Forbidden, discord.NotFound):
            pass

    async def _reply(self,
                     ctx: commands.Context,
                     *,
                     title: str,
                     description: str,
                     level: str,
                     show_footer: bool = True) -> discord.Message | None:
        await self._delete_command(ctx)

        embed = make_embed(title=title, description=description, level=level)

        if show_footer:
            embed.set_footer(text=f"Action by : {ctx.author}",
                             icon_url=ctx.author.display_avatar.url)

        return await ctx.channel.send(embed=embed)

    async def cog_check(self, ctx: commands.Context) -> bool:
        if ctx.guild is None:
            return True
        if not isinstance(ctx.author, discord.Member):
            return False
        if ctx.author.id == ctx.guild.owner_id:
            return True

        perms = ctx.author.guild_permissions
        if perms.administrator or perms.manage_nicknames:
            return True

        return await super().cog_check(ctx)

    def _normalize(self, text: str) -> str:
        return unicodedata.normalize("NFKC", text)

    def _clean_text(self, text: str) -> str:
        return "".join(c for c in text.lower() if c.isalnum())

    def _contains_profanity(self, text: str) -> bool:
        clean = self._clean_text(text)
        return any(word in clean for word in PROFANITY_BLOCKLIST)

    def _bot_can_modify(self, guild: discord.Guild,
                        target: discord.Member) -> bool:
        if self.bot.user is None:
            return False

        bot_member = guild.get_member(self.bot.user.id)
        if not bot_member:
            return False

        return (bot_member.guild_permissions.manage_nicknames
                and target != guild.owner
                and target.top_role < bot_member.top_role)

    def _moderator_can_modify(self, guild: discord.Guild,
                              moderator: discord.Member,
                              target: discord.Member) -> bool:
        if moderator.id == guild.owner_id:
            return True
        if target == guild.owner:
            return False
        if target == moderator:
            return True

        return target.top_role < moderator.top_role

    @commands.command(name="rename", aliases=["nick", "setnick"])
    @commands.guild_only()
    @commands.cooldown(COOLDOWN_RATE, COOLDOWN_PER, commands.BucketType.user)
    async def rename(self,
                     ctx: commands.Context,
                     *,
                     args: str | None = None) -> None:
        guild = ctx.guild
        if guild is None:
            return

        if not isinstance(ctx.author, discord.Member):
            return

        moderator: discord.Member = ctx.author
        prefix = ctx.clean_prefix
        invoked_with = ctx.invoked_with

        if self.bot.user is None:
            return

        bot_member = guild.get_member(self.bot.user.id)
        if not bot_member or not bot_member.guild_permissions.manage_nicknames:
            await self._reply(
                ctx,
                title="Missing Permissions",
                description="I need Manage Nicknames permission.",
                level="ERROR")
            return

        if not args:
            await self._reply(
                ctx,
                title="Missing Nickname",
                description=("Usage:\n"
                             f"`{prefix}{invoked_with} <nickname>`\n"
                             f"`{prefix}{invoked_with} @user <nickname>`\n"
                             f"`{prefix}{invoked_with} reset`"),
                level="WARNING")
            return

        target: discord.Member
        nickname: str

        if ctx.message.mentions:
            raw = ctx.message.mentions[0]
            member = guild.get_member(raw.id)
            if not member:
                await self._reply(ctx,
                                  title="Invalid Target",
                                  description="User must be in this server.",
                                  level="ERROR")
                return
            target = member
            nickname = args.replace(raw.mention, "", 1).strip()
        else:
            target = moderator
            nickname = args.strip()

        if not nickname:
            await self._reply(ctx,
                              title="Missing Nickname",
                              description="Please provide a nickname.",
                              level="ERROR")
            return

        is_self = target == moderator
        if not is_self:
            if not self._moderator_can_modify(guild, moderator, target):
                await self._reply(ctx,
                                  title="Permission Denied",
                                  description="You cannot modify this user.",
                                  level="ERROR")
                return

        if not self._bot_can_modify(guild, target):
            await self._reply(
                ctx,
                title="Role Hierarchy Issue",
                description="My role is too low to modify this user.",
                level="ERROR")
            return

        if nickname.lower() == "reset":
            if target.nick is None:
                await self._reply(
                    ctx,
                    title="No Nickname Set",
                    description="That user does not have a nickname.",
                    level="INFO")
                return

            old_nick = target.display_name

            try:
                await target.edit(nick=None,
                                  reason=f"Nickname reset by {moderator}")
            except discord.Forbidden:
                await self._reply(ctx,
                                  title="Missing Permissions",
                                  description="I cannot change this nickname.",
                                  level="ERROR")
                return
            except discord.HTTPException:
                await self._reply(ctx,
                                  title="Discord Error",
                                  description="Failed to update nickname.",
                                  level="ERROR")
                return

            success_emoji = EMOJIS.get("success") or "✅"
            await self._reply(
                ctx,
                title="Nickname Reset",
                description=
                f"{success_emoji} Nickname removed for {target.mention}.",
                level="SUCCESS")

            try:
                await send_mod_log(
                    guild=guild,
                    category="MODERATION",
                    title="Nickname Reset",
                    description=
                    f"{moderator} reset nickname of {target.mention}",
                    level="INFO",
                    actor=moderator,
                    target=target,
                    extra_fields={"Old Nickname": old_nick})
            except Exception:
                logger.exception("Failed to send rename moderation log")

            return

        # RENAME
        nickname = self._normalize(nickname)
        nickname = re.sub(r"[\u200B-\u200D\uFEFF]", "", nickname)
        nickname = " ".join(nickname.split())[:32]

        if "@everyone" in nickname or "@here" in nickname:
            await self._reply(ctx,
                              title="Invalid Nickname",
                              description="Mass mentions are not allowed.",
                              level="ERROR")
            return

        if self._contains_profanity(nickname):
            await self._reply(
                ctx,
                title="Blocked Nickname",
                description="Nickname contains prohibited content.",
                level="ERROR")
            return

        old_nick = target.display_name

        try:
            await target.edit(nick=nickname,
                              reason=f"Nickname changed by {moderator}")
        except discord.Forbidden:
            await self._reply(ctx,
                              title="Missing Permissions",
                              description="I cannot change this nickname.",
                              level="ERROR")
            return
        except discord.HTTPException:
            await self._reply(ctx,
                              title="Discord Error",
                              description="Failed to update nickname.",
                              level="ERROR")
            return

        success_emoji = EMOJIS.get("success") or "✅"
        await self._reply(
            ctx,
            title="Nickname Updated",
            description=
            f"{success_emoji} {target.mention} is now **{nickname}**.",
            level="SUCCESS")

        try:
            await send_mod_log(
                guild=guild,
                category="MODERATION",
                title="Nickname Changed",
                description=f"{moderator} changed nickname of {target.mention}",
                level="INFO",
                actor=moderator,
                target=target,
                extra_fields={
                    "Old Nickname": old_nick,
                    "New Nickname": nickname,
                })
        except Exception:
            logger.exception("Failed to send rename moderation log")


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(RenameSystem(bot))

File: src/commands/admin/rename.js
import { PermissionFlagsBits, SlashCommandBuilder } from "discord.js";
import { createCommand } from "../../core/command.js";
import { makeEmbed } from "../../core/embeds.js";
import { EMOJIS } from "../../core/emojis.js";
import { sendModLog } from "../../utils/modLog.js";

const slashBuilder = new SlashCommandBuilder()
  .setName("rename")
  .setDescription("Change or reset a member's server nickname")
  .addUserOption((opt) => opt.setName("user").setDescription("The target member").setRequired(true))
  .addStringOption((opt) => opt.setName("nickname").setDescription("The new nickname or 'reset' to clear").setRequired(true));

export default createCommand({
  name: "rename",
  description: "Change or reset a member's server nickname",
  category: "Admin",
  aliases: ["nick", "setnick"],
  modOnly: true,
  requiredPermission: PermissionFlagsBits.ManageNicknames,
  slashBuilder,

  async execute(ctx) {
    const { guild, user, member: moderator } = ctx;
    if (!guild) return;

    const botMember = guild.members.me;
    if (!botMember?.permissions?.has(PermissionFlagsBits.ManageNicknames)) {
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Missing Permissions",
            description: `${EMOJIS.get("fail") || "❌"} I need the **Manage Nicknames** permission to change nicknames.`,
            level: "ERROR",
          }),
        ],
        ephemeral: true,
      });
    }

    let targetUserId = ctx.options.user;
    let newNick = ctx.options.nickname;

    if (!targetUserId) {
      // Prefix handling
      if (ctx.message?.mentions?.members?.first()) {
        const mentioned = ctx.message.mentions.members.first();
        targetUserId = mentioned.id;
        newNick = ctx.options._args?.slice(1).join(" ");
      } else if (ctx.options._args?.length > 1 && /^\d{17,20}$/.test(ctx.options._args[0])) {
        targetUserId = ctx.options._args[0];
        newNick = ctx.options._args.slice(1).join(" ");
      } else if (ctx.options._args?.length > 0) {
        // Renaming self if no user mentioned
        targetUserId = user.id;
        newNick = ctx.options._args.join(" ");
      }
    }

    if (!targetUserId || !newNick) {
      const prefix = ctx.client?.config?.PREFIX || "!";
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Missing Arguments",
            description: `Usage:\n• \`${prefix}rename @user <nickname>\`\n• \`${prefix}rename @user reset\`\n• \`/rename user: @user nickname: <new nickname>\``,
            level: "WARNING",
          }),
        ],
        ephemeral: true,
      });
    }

    const targetMember = await guild.members.fetch(targetUserId).catch(() => null);
    if (!targetMember) {
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Member Not Found",
            description: `${EMOJIS.get("fail") || "❌"} The specified member is not in this server.`,
            level: "ERROR",
          }),
        ],
        ephemeral: true,
      });
    }

    // Role hierarchy checks
    const isOwner = guild.ownerId === user.id;
    if (targetMember.id === guild.ownerId && !isOwner) {
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Permission Denied",
            description: `${EMOJIS.get("fail") || "❌"} You cannot change the nickname of the server owner.`,
            level: "ERROR",
          }),
        ],
        ephemeral: true,
      });
    }

    if (moderator && targetMember.id !== user.id && !isOwner) {
      if (targetMember.roles.highest.position >= moderator.roles.highest.position) {
        return await ctx.reply({
          embeds: [
            makeEmbed({
              title: "Permission Denied",
              description: `${EMOJIS.get("fail") || "❌"} You cannot rename someone with an equal or higher role than yourself.`,
              level: "ERROR",
            }),
          ],
          ephemeral: true,
        });
      }
    }

    if (targetMember.id !== botMember.id && targetMember.roles.highest.position >= botMember.roles.highest.position) {
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Role Hierarchy Issue",
            description: `${EMOJIS.get("fail") || "❌"} My role is not high enough to modify ${targetMember}'s nickname.`,
            level: "ERROR",
          }),
        ],
        ephemeral: true,
      });
    }

    const oldNick = targetMember.displayName;

    // Reset flow
    if (newNick.toLowerCase() === "reset") {
      if (!targetMember.nickname) {
        return await ctx.reply({
          embeds: [
            makeEmbed({
              title: "No Nickname Set",
              description: `${EMOJIS.get("warning") || "ℹ️"} ${targetMember} does not have an active custom nickname.`,
              level: "INFO",
            }),
          ],
        });
      }

      try {
        await targetMember.setNickname(null, `Nickname reset by ${user.tag || user.username}`);
      } catch (err) {
        return await ctx.reply({
          embeds: [
            makeEmbed({
              title: "Update Failed",
              description: `${EMOJIS.get("fail") || "❌"} Failed to reset nickname: ${err?.message || "Discord API error"}.`,
              level: "ERROR",
            }),
          ],
          ephemeral: true,
        });
      }

      await sendModLog({
        guild,
        category: "MODERATION",
        title: "Nickname Reset",
        description: `<@${user.id}> reset the nickname of <@${targetMember.id}>.`,
        level: "INFO",
        actor: user,
        target: targetMember.user,
        extraFields: { "Old Nickname": oldNick },
      });

      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Nickname Reset",
            description: `${EMOJIS.get("success") || "✅"} Nickname reset successfully for ${targetMember}.`,
            level: "SUCCESS",
          }),
        ],
      });
    }

    // Sanitize nickname
    const cleaned = newNick.replace(/[\u200B-\u200D\uFEFF]/g, "").trim();
    if (cleaned.includes("@everyone") || cleaned.includes("@here")) {
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Invalid Nickname",
            description: `${EMOJIS.get("fail") || "❌"} Mass mentions (\`@everyone\`, \`@here\`) are not allowed in nicknames.`,
            level: "ERROR",
          }),
        ],
        ephemeral: true,
      });
    }

    const clampedNick = cleaned.slice(0, 32);

    try {
      await targetMember.setNickname(clampedNick, `Nickname changed by ${user.tag || user.username}`);
    } catch (err) {
      return await ctx.reply({
        embeds: [
          makeEmbed({
            title: "Update Failed",
            description: `${EMOJIS.get("fail") || "❌"} Failed to change nickname: ${err?.message || "Discord API error"}.`,
            level: "ERROR",
          }),
        ],
        ephemeral: true,
      });
    }

    await sendModLog({
      guild,
      category: "MODERATION",
      title: "Nickname Changed",
      description: `<@${user.id}> changed the nickname of <@${targetMember.id}> to **${clampedNick}**.`,
      level: "INFO",
      actor: user,
      target: targetMember.user,
      extraFields: {
        "Old Nickname": oldNick,
        "New Nickname": clampedNick,
      },
    });

    return await ctx.reply({
      embeds: [
        makeEmbed({
          title: "Nickname Updated",
          description: `${EMOJIS.get("success") || "✅"} ${targetMember} is now **${clampedNick}**.`,
          level: "SUCCESS",
        }),
      ],
    });
  },
});
